### PR TITLE
Additoinal losses

### PR DESCRIPTION
**Pull request by Jyun1998**
## Proposed changes

Added commonly used losses with tests

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the necessary documentation (if needed)

### CODE REVIEW

**Review by awni — CHANGES_REQUESTED** [review]
Thanks for the contribution!! 

We already have a `hinge` and `huber` loss already. The `focal`, `dice`, and `contrastive` are in #324. Please remove all of them and rebase to resolve conflicts.

If you want to add the `cosine` loss in this PR that would be great.

**Comment by Jyun1998** [comment]
Hi @awni , rebased and make cos loss to remain on this pr. 

Also commented on #324 based on my implementation.

Thanks :)

**Review by awni — CHANGES_REQUESTED** [review]
Thanks! I left a couple  comments, please address then let me know when ready for re-review.

File: python/tests/test_nn.py [hunk 1]
```diff
@@ -843,3 +869,4 @@ def test_dropout3d(self):
 
 if __name__ == "__main__":
     unittest.main()
+
```
[the remark this answers is not in the document]
Comment (Jyun1998) [reply python/tests/test_nn.py]: updated :)

**Comment by awni** [comment]
Hey @Jyun1998 sorry but somehow my main comment did not get included. I must have not pushed the save button by accident.

Basically I'm wondering what reference you use for this loss as it looks quite different than the similarly named [cosine similarity loss in PyTorch](https://pytorch.org/docs/stable/generated/torch.nn.CosineSimilarity.html) ?

For example there is a `target` and a `margin` both of which I do not expect.  Unless there is a good reason for the difference, I would suggest we follow the PyTorch implementation as a reference.

**Comment by awni** [comment]
@Jyun1998 are you still planning to follow up on this?

**Comment by Jyun1998** [comment]
> @Jyun1998 are you still planning to follow up on this?

https://github.com/keras-team/keras/blob/v2.14.0/keras/losses.py#L1162-L1236

Hi awni, according to common tensorflow and pytorch implementaiton, the functions are composed of doing l2 norm to each embedding and returns the negative of dot product of both embedding.

My codes also do so, and margin-based loss is applying if it's necessary :)

**Comment by awni** [comment]
@Jyun1998 got it. We should keep it simple until we see that we need more features. Could you follow the [PyTorch cosine similarity](https://pytorch.org/docs/stable/generated/torch.nn.functional.cosine_similarity.html) loss? I think that one covers the most common case (margin / targets are are niche and possibly never needed so we don't want to add them to the API until we are sure they are necessary)

**Comment by Jyun1998** [comment]
> @Jyun1998 got it. We should keep it simple until we see that we need more features. Could you follow the [PyTorch cosine similarity](https://pytorch.org/docs/stable/generated/torch.nn.functional.cosine_similarity.html) loss? I think that one covers the most common case (margin / targets are are niche and possibly never needed so we don't want to add them to the API until we are sure they are necessary)

I also agree. Even though there's also margin for pytorch F.cosine_similarity, it defaultly do not use it. 

There's also slight change needed for changes and I will test the function and commit asap. Thanks for the review

**Comment by awni** [comment]
> Even though there's also margin for pytorch F.cosine_similarity

I don't see the margin in the docs? Is it in the source code?

Screenshot 2024-01-07 at 6 30 32 AM

**Comment by Jyun1998** [comment]
> > Even though there's also margin for pytorch F.cosine_similarity
> 
> I don't see the margin in the docs? Is it in the source code?
> 
> Screenshot 2024-01-07 at 6 30 32 AM

https://pytorch.org/docs/stable/generated/torch.nn.CosineEmbeddingLoss.html#torch.nn.CosineEmbeddingLoss

**Comment by awni** [comment]
I see, thanks. Yes let's go with the plain cosine similarity for now. Thank you!

**Comment by awni** [comment]
Also could you rebase and resolve conflicts?

**Comment by Jyun1998** [comment]
> Also could you rebase and resolve conflicts?

Am I correct that losses test codes are gone?

**Comment by Jyun1998** [comment]
> Also could you rebase and resolve conflicts?

Could you check? Thanks :)

File: python/mlx/nn/losses.py [hunk 2]
```diff
@@ -372,3 +372,45 @@ def log_cosh_loss(
...
+
+    .. math::
+
+        \text{cosine_similiarity_loss} = \frac{\sum_{i} e_{1,i} \cdot e_{2,i}}{\max(\|e_1\|, \varepsilon) \cdot \max(\|e_2\|, \varepsilon)}
+
+
+    Args:
+        embeddings1 (mx.array): Embeddings for the first set of samples.
+        embeddings2 (mx.array): Embeddings for the second set of samples.
+        targets (mx.array): The target values (cosine similarity between embeddings).
```
Comment (awni) [inline python/mlx/nn/losses.py:402]: I would switch to use `embeddings1` and `embeddings` and remove `targets`

File: python/mlx/nn/losses.py [hunk 3]
```diff
@@ -372,3 +372,45 @@ def log_cosh_loss(
...
+        margin (float, optional): Margin for dissimilar pairs. Default: ``0.0``.
+        reduction (str, optional): Specifies the reduction to apply to the output:
+          ``'none'`` | ``'mean'`` | ``'sum'``. Default: ``'none'``.
+
+    Returns:
+        mx.array: The computed Cosine Similarity loss.
+    """
+    inputs_norm = mx.maximum(mx.sqrt(mx.sum(mx.square(inputs), axis=1)), eps)
+    targets_norm = mx.maximum(mx.sqrt(mx.sum(mx.square(targets), axis=1)), eps)
```
Comment (awni) [inline python/mlx/nn/losses.py:411]: Nit you could do `mx.linalg.norm(embeddings)` here instead of the `sqrt` + `sum` + `square`.

File: python/mlx/nn/losses.py [hunk 4]
```diff
@@ -372,3 +372,45 @@ def log_cosh_loss(
     loss = mx.logaddexp(errors, -errors) - math.log(2)
 
     return _reduce(loss, reduction)
+
+
+def cosine_similarity_loss(
+    inputs: mx.array,
+    targets: mx.array,
+    eps: float = 1e-8,
+    reduction: str = "none",
```
Comment (awni) [inline python/mlx/nn/losses.py:381]: Could you make the `axis` a parameter to this function instead of hardcoding it?

File: python/mlx/nn/losses.py [hunk 5]
```diff
@@ -372,3 +372,45 @@ def log_cosh_loss(
...
+    r"""
+    Computes the Cosine Similarity loss between inputs and targets.
+
+    Note that inputs must be between -1 and 1. When it is a negative number
+    between -1 and 0, 0 indicates orthogonality and values closer to -1 indicate
+    greater similarity. The values closer to 1 indicate greater dissimilarity.
+    This makes it usable as a loss function in a setting where you try to
+    maximize the proximity between predictions and targets. If either
+    `inputs` or `targets` is a zero vector, cosine similarity will be 0
+    regardless of the proximity between predictions and targets.
```
Comment (awni) [inline python/mlx/nn/losses.py:392]: Why do the inputs need to be between `-1` and `1`? I think you can remove this paragraph.

File: python/mlx/nn/losses.py [hunk 6]
```diff
@@ -372,3 +372,45 @@ def log_cosh_loss(
...
+
+    .. math::
+
+        \text{cosine_similiarity_loss} = \frac{\sum_{i} e_{1,i} \cdot e_{2,i}}{\max(\|e_1\|, \varepsilon) \cdot \max(\|e_2\|, \varepsilon)}
```
Comment (awni) [inline python/mlx/nn/losses.py:396]:
```suggestion
     The cosine similarity loss is given by
     
        .. math::

        \frac{x_1 \cdot x_2}{\max(\|x_1\|  \cdot \|x_2\|, \repsilon)}
```

File: python/mlx/nn/losses.py (hunk 3 above)
Comment (awni) [inline python/mlx/nn/losses.py:411]: Use one maximum on the denominator instead of two. So you could move the maximum below to L415 on the denominator.

File: python/tests/test_losses.py [hunk 7]
```diff
@@ -274,6 +274,25 @@ def test_log_cosh_loss(self):
...
+
+        # Test with reduction 'mean'
+        losses_mean = nn.losses.triplet_loss(inputs, targets, reduction="mean")
+        expected_mean = mx.mean(expected_none)
+        self.assertTrue(mx.allclose(losses_mean, expected_mean))
+
+        # Test with reduction 'sum'
+        losses_sum = nn.losses.triplet_loss(inputs, targets, reduction="sum")
+        expected_sum = mx.sum(expected_none)
+        self.assertTrue(mx.allclose(losses_sum, expected_sum))
```
Comment (awni) [inline python/tests/test_losses.py:294]: You are testing the triple_loss here?

**Review by awni — CHANGES_REQUESTED** [review]
A few more changes please.

File: python/mlx/nn/losses.py (hunk 3 above)
Comment (Jyun1998) [reply python/mlx/nn/losses.py]: Yes, I've tried it and `mx.linalg.norm(embeddings1, axis=axis)` have problem with `axis=1` for `not supporting argument type` where it should allow int val. 

Know any reasons behind?

Comment (awni) [reply python/mlx/nn/losses.py]: Can you give me a snippet of code which reproduces that, the axis param works fine for me...

```
>>> mx.linalg.norm(mx.zeros((10,)), axis=0)
array(0, dtype=float32)
```

**Review by awni — APPROVED** [review]
Looks great, thanks for the contribution @Jyun1998 !
